vulcanai/models/basenetwork.py

```diff
@@ -494,6 +538,59 @@ def _assert_same_devices(self, comparison_device=None):
                     comparison_device,
                     incompatible_collector))
 
+    # modified from
+    # https://github.com/Bjarten/early-stopping-pytorch/blob/master
+    # /pytorchtools.py
+    class EarlyStopping:
+        """Early stops the training if validation loss doesn't improve after
+        a given patience."""
+
+        def __init__(self, patience=2, verbose=False):
+            """
+            Args:
+                patience (int): How long to wait after last time validation
+                loss improved.
+                                Default: 7
+                verbose (bool): If True, prints a message for each
+                validation loss improvement.
+                                Default: False
+            """
+            self.patience = patience
+            self.verbose = verbose
+            self.counter = 0
+            self.best_score = None
+            self.early_stop = False
+            self.val_loss_min = np.Inf
+            self.best_model = None
+            self.save_path = ""
+
+        def __call__(self, score, model):
+
+            if self.best_score is None:
+                self.best_score = score
+                self.save_checkpoint(score, model)
+            elif score < self.best_score:
+                self.counter += 1
+                logger.info('EarlyStopping counter: {} out of {}'.format(
+                    self.counter, self.patience
+                ))
+                if self.counter >= self.patience:
+                    self.early_stop = True
+
+            else:
+                self.best_score = score
+                self.save_checkpoint(score, model)
+                self.counter = 0
+
+        def save_checkpoint(self, val_loss, model):
+            '''Saves model when validation loss decrease.'''
+            if self.verbose:
+                logger.info(
+                    f'Validation loss decreased ({self.val_loss_min:.6f} --> '
+                    f'{val_loss:.6f}).  Saving model ...')
+            self.save_path = model.save_model()
```

Comment:
is there a specific reason that the early stopping checkpoint is saved and not just stored in memory? I guess it's better incase the script suddenly stops but wondering what your thoughts were on this.

Reply:
yeah that's why I did it, although a downside is slightly increased latency